New (breaking) Traits based release

evetion opened the pull request:
Based on https://github.com/JuliaGeo/GeoInterfaceRFC.jl

- [x] Decide what to do about `primitives.jl` (removed it)
- Update all integrations in other packages (https://github.com/JuliaGeo/GeoInterfaceRFC.jl/blob/master/INTEGRATIONS.md)
	- [ ] ArchGDAL (https://github.com/yeesian/ArchGDAL.jl/pull/290)
	- [ ] LibGEOS (blocked by https://github.com/JuliaGeo/LibGEOS.jl/issues/101 and https://github.com/JuliaGeo/LibGEOS.jl/issues/96)
	- [ ] GeometryBasics ([commented on rewrite](https://github.com/JuliaGeometry/GeometryBasics.jl/pull/173#issuecomment-1108174450))
	- [ ] ShapeFile (https://github.com/JuliaGeo/Shapefile.jl/pull/72)
	- [ ] Rasters
	- [ ] GeoDataFrames
	- [ ] Meshes.jl
	- [ ] FlatGeobuf.jl (https://github.com/evetion/FlatGeobuf.jl)
	- [ ] GeoParquet.jl (depends on WellKnownGeometry) 
	- [x] WellKnownGeometry.jl (https://github.com/evetion/WellKnownGeometry.jl)
	- [x] GeoJSONTables (https://github.com/visr/GeoJSONTables.jl)

@visr @yeesian @juliohm

juliohm commented on src/GeoInterfaceRFC.jl:
```diff
@@ -0,0 +1,17 @@
+module GeoInterfaceRFC
```
The module name needs to be updated in the source and test files :+1:

rafaqz commented:
I've been reading over this too now I'm using GeoInterface more. It seems like a real improvement, OI'm keen to help getting it merged.

My comments so far are:

- it seems that `extent` replaces `bbox` ? I can't find the implementation of `extent`
- `crs`  should probably return `nothing` instead of `missing` by default. `nothing` is more often used like that.
- We will probably need more fallbacks that return `nothing` when called on an object that doesn't support the trait. I've just added these to the old GeoInterface for `bbox` and `crs` so we can call it just in case, as e.g. Shapefiles have `bbox` precalculated for everything, but other gemoetries may not.
- We should make sure that `coordnames` and `x`/`y`constant-propagates and findfist` isn't a runtime lookup.

rafaqz commented:
I would like to propose an `Extent` object to be used instead of `Rectangle` for the return value of `extent`. It wraps a NamedTuple because all fields are not always present.

```julia
struct Extent{T<:NamedTuple}
    bounds::T
end
Extent(; kw...) = Extent(values(kw))
Base.getproperty(ext::Extent, x::Symbol) = getproperty(getfield(ext, :bounds), x)
ext = Extent(X=(0.0, 10.0), Y=(0.0, 10.0))
ext.X
```

Giving it a specific type means we can dispatch on it, and including the dimension names means we know their order in the object.

evetion commented:
Ah yes, I agree that we need some sort of bounds object. I've had issues finding a generic one, hence the default to a `Rectangle`. Nowadays I just revert to using a namedtuple, but am frustrated by the fact that a `(min_x=1, min_y=2) != (min_y=2, min_x=1)`(nor its types), which for all purposes would be equal.

@visr Do you remember my magical bbox code that you reviewed and we shelved for it being too complex? I can't seem to find it.

rafaqz commented:
Yeah, a `NamedTuple` is not enough. But its close to it... that's why I've just wrapped it with `Extent`. You get a lot of the code in your `BoundingBox` object for free just by forwarding to `NamedTuple`. 

With the `Extent` wrapper we could define `==` to check that both objects have the same dimension names (but not necessarily the same order), and then match all of the values by the names.

rafaqz commented:
Better implementation including `==`

```julia
struct Extent{T<:NamedTuple}
    bounds::T
end
Extent(; kw...) = Extent(values(kw))

bounds(ext::Extent) = getfield(ext, :bounds)

Base.getproperty(ext::Extent, x::Symbol) = getproperty(bounds(ext), x)
Base.getindex(ext::Extent, x::Symbol) = getindex(bounds(ext), x)
import Base: ==
function (==)(a::Extent{<:NamedTuple{K1}}, b::Extent{<:NamedTuple{K2}}) where {K1, K2}
    length(K1) == length(K2) || return false
    all(map(k -> k in K1, K2)) || return false
    return all(map((k -> a[k] == b[k]), K1))
end
```

And the order doesn't matter for `==`:
```julia
julia> Extent(X=(0, 1), Y=(3.0, 4.0)) == Extent(Y=(3.0, 4.0), X=(0, 1))
true
```

rafaqz commented:
Cautionary tail against using `missing` as a default return value in an interface:
https://github.com/JuliaArrays/ArrayInterface.jl/pull/245

evetion commented:
Thanks, that's good to know when updating this PR.

evetion commented:
Some updates:
- Removed merge conflicts
- Added documentation (still more to do)
- Removed primitives, they could be added back (Vector/Tuple is a Point? but what about other geometries?)
- Added `isgeometry` check, mimicking `istable` from Tables.jl
- Enumerated current packages using GeoInterface

> * it seems that `extent` replaces `bbox` ? I can't find the implementation of `extent`

Will now use Extents.jl once registered.
> * `crs`  should probably return `nothing` instead of `missing` by default. `nothing` is more often used like that.

Done

---
TODO
> * We will probably need more fallbacks that return `nothing` when called on an object that doesn't support the trait. I've just added these to the old GeoInterface for `bbox` and `crs` so we can call it just in case, as e.g. Shapefiles have `bbox` precalculated for everything, but other gemoetries may not.
> * We should make sure that `coordnames` and `x`/`y`constant-propagates and `findfist` isn't a runtime lookup.

And 
- [ ] Implement this in other packages (at least the most used)
- [ ] Think about automatic `convert`s between geometries based on GeoInterface
- [ ] Do we need a native WKT/WKB implementation?

evetion commented on src/defaults.jl:
```diff
@@ -0,0 +1,69 @@
+
+coordnames(::AbstractGeometry, geom) = default_coord_names[1:ncoord(geom)]
+
+# Maybe hardcode dimension order? At least for X and Y?
```
In the SF spec, you have either, XY, XYZ, XYM or XYZM. No other orders are "possible". You see this in the Shapefile structs as well. In that sense we could hardcode the XY as always being the first two dimensions, but we might rather leave this open to more names/orders for some freedom?

rafaqz replied:
Yeah, better to leave it open.

evetion commented:
Based on my PR for ArchGDAL, I noticed that we don't have a `coordinates` anymore. I've implemented one for backwards compatibility, also for compatibility the Python/GeoJSON spec. It is now implemented as a fallback, see [here](https://github.com/JuliaGeo/GeoInterface.jl/pull/33/commits/2ae95813d020073afb85ee4a6b827fedfe40b6ab#diff-e7bdce3e18fa78e2af458c94e3a44028dfb7a7211de96fe0d7ce0321040197daR71), which works pretty nice and seems to compile ok.

visr commented:
Thanks for pushing this forward! I agree that actually implementing the interface for several packages and trying it out (e.g. converting a geometry through the interface, compared to doing so with both packages available) will give a good idea of the status of the PR.

It's currently not in the top post, but GeometryBasics integration will be a key one here. Way back I started a branch here: https://github.com/visr/GeometryBasics.jl/tree/rfc, but it would be good to get feedback from the GeometryBasics authors, also in light of https://github.com/JuliaGeometry/GeometryBasics.jl/pull/173.

evetion commented:
What do we prefer in terms of function signatures

```julia
geomtype(geom) = Point
function(::Type{Point}, geom)
```

```julia
geomtype(geom) = Point()
function(::Point}, geom)
```

rafaqz commented:
Using objects is better for type stability if you ever have to pass them around.

But it seems confusing to me that a constructed `Point()` is an empty singleton trait and not an actual point object.

evetion commented:
> Using objects is better for type stability if you ever have to pass them around.

Ok, reverted my changes.

> Maybe we should somehow suggest that its a trait in the name?

I've suffixed all types with `Trait`. So `Point` -> `PointTrait`. If we do not like this, what about `PointLike`? 😉

evetion commented:
@visr @rafaqz 

To be fully backwards compatible with the previous GeoInterface, shall we add 
```
    geometry(::AbstractFeature)  # old
    properties(::AbstractFeature)  # old
	isfeature(x) # new
```

as well? Most geometries don't come alone, but in Features (rows).

evetion commented on docs/src/background/sf.md:
```diff
@@ -0,0 +1,60 @@
+We generalized [`ngeom`](@ref) and [`getgeom`](@ref) to apply to 
+all geometries, not just a [`AbstractGeometryCollection`](@ref)s.
+
+We also simplified the dimension functions. From the three original (`dimension`, `coordinateDimension`, `spatialDimension`) there's now only the coordinate dimension, so not to overlap with the Julia `ndims`.
```
I've clarified the use of these functions. We only keep ncoords. The others are _topological_ dimensions, which don't relate to any Julia concept.

evetion commented:
@rafaqz @visr @yeesian @juliohm @joshday I think this PR is feature complete. I will only improve the test coverage, but otherwise plan to merge it and release v1.0 come next weekend. 🎉 

It would be great if you could give it a last review for any blockers.

rafaqz commented:
I just hit a `findfirst` error here - it doesn't work on `Tuple`. 

```julia
const default_coord_names = (:X, :Y, :Z, :M)

x(::AbstractPointTrait, geom) = getcoord(geom, findfirst(coordnames(geom), :X))
y(::AbstractPointTrait, geom) = getcoord(geom, findfirst(coordnames(geom), :Y))
z(::AbstractPointTrait, geom) = getcoord(geom, findfirst(coordnames(geom), :Z))
m(::AbstractPointTrait, geom) = getcoord(geom, findfirst(coordnames(geom), :M))
```

evetion commented:
> I just hit a `findfirst` error here - it doesn't work on `Tuple`.

My bad, I've fixed it and added a test for it. Will add some more tests on the defaults, and enable coverage to prevent this.

rafaqz commented on .github/workflows/CI.yml:
```diff
@@ -13,13 +18,11 @@ jobs:
-          - '1'
-          - 'nightly'
+          - "1"
+          - "1.7"
```
```suggestion
          - "1.6"
```

Shouldn't we test on the LTS version?
